x/stake/endblock.go

```diff
@@ -252,3 +265,63 @@ func handleMatureUnbondingDelegations(k keeper.Keeper, ctx sdk.Context) ([]types
 
 	return completed, events
 }
+
+const (
+	maxProcessedRefundCount = 50
+)
+
+func handleRefundCrossStake(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	iterator := k.IteratorAllDelegations(ctx)
+	defer iterator.Close()
+	var refundEvents sdk.Events
+	count := 0
+	for ; iterator.Valid(); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.CDC(), iterator.Key(), iterator.Value())
+		if delegation.CrossStake {
+			result := handleMsgSideChainUndelegate(ctx, types.MsgSideChainUndelegate{
+				DelegatorAddr: delegation.DelegatorAddr,
+				ValidatorAddr: delegation.ValidatorAddr,
+				Amount:        sdk.NewCoin(k.BondDenom(ctx), delegation.GetShares().RawInt()),
+				SideChainId:   k.ScKeeper.BscSideChainId(ctx),
+			}, k)
+			refundEvents = refundEvents.AppendEvents(result.Events)
+
+			count++
+		}
+		if count >= maxProcessedRefundCount {
+			break
+		}
+	}
+	return refundEvents
+}
+
+func closeSideChainChannels(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	var events sdk.Events
+	if sdk.IsUpgrade(sdk.BCFusionSecondHardFork) {
+		sideChainId := k.ScKeeper.BscSideChainId(ctx)
+		vp := k.GetSideChainTotalVotingPower(ctx, sideChainId)
+		if vp.LTE(sdk.NewDecFromInt(5_000_000)) {
+			// disable side chain channels
+			id := k.ScKeeper.Config().DestChainNameToID(sideChainId)
+
+			for _, channelId := range k.ScKeeper.Config().ChannelIDs() {
```

Comment:
We can not close channel in random order, the `gov` channel have to be the last one, becase the underlying channle of these package are `gov` channel. 

Besides `SaveChannelSettingChangeToIbc`, you have to disable the BC channels as well

Reply:
fixed.